lib/ansible/modules/cloud/azure/azure_rm_virtualmachine_facts.py

```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018 Gustavo Muniz do Carmo <gustavo@esign.com.br>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_virtualmachine_facts
+
+version_added: "2.6"
+
+short_description: Get virtual machine facts.
+
+description:
+  - Get facts for all virtual machines of a resource group.
+
+options:
+  resource_group:
+    description:
+      - Name of the resource group containing the virtual machines.
+  name:
+    description:
+      - Name of the virtual machine.
+  tags:
+    description:
+      - Limit results by providing a list of tags. Format tags as 'key' or 'key:value'.
+
+extends_documentation_fragment:
+  - azure
+
+author:
+  - "Gustavo Muniz do Carmo (@gustavomcarmo)"
+
+'''
+
+EXAMPLES = '''
+  - name: Get facts for all virtual machines of a resource group
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+
+  - name: Get facts by name
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+      name: vm
+
+  - name: Get facts by tags
+    azure_rm_virtualmachine_facts:
+      resource_group: Testing
+      tags:
+        - testing
+        - foo:bar
+'''
+
+RETURN = '''
+azure_virtualmachines:
```

Comment:
I think we need to take a closer step for this and decide how we want to move forward. There are 2 things we should discuss in the wider group;

* The values are currently in camelCase and not snake_case, we should try and use the snake variable wherever possible. Luckily we have an easy to use function that can do this for you https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L376
* The values are quite deeply nested and sometimes in keys that don't really make sense, i.e. the `properties` key that seems to be on each element

Having these values deeply nested makes things like getting the public IP address quite a cumbersone process, e.g. `azure_virtualmachines[0].networkProfile.networkInterfaces[0].properties.ipConfigurations[0].properties.publicIPAddress.properties.ipAddress`. Compare this to [ec2_instance_facts](http://docs.ansible.com/ansible/devel/modules/ec2_instance_facts_module.html#ec2-instance-facts-module) I only need to do `instances.public_ip_address` which is a lot easier to read and write.

We don't need to conform to the same structure that the ec2 facts module is but it is definitely something we should try and keep similar. I think we should discuss this as a group in the next Azure Working Group meeting and can be added to the agenda for the meetings [here](https://github.com/ansible/community/issues/281)

Reply:
@jborean93 we had discussion with @nitzmahone on this recently. Actually we agreed that we will probably should have a few options for facts modules:
- raw - whatever is returned by REST API
- default - should match azure_rm_xxx input
- minimal - azure_rm_xxx minimal output
So i am actually planning to upgrade all Azure modules like this.
We could do it with this module now as well, but perhaps in next PR.

@gustavomcarmo could you make such changes? or would you like me to do appropriate updates and then merge?

Reply:
@jborean93 in the same module. just have an option.

Reply:
@jborean93 Agree about snake_case too. @zikalino I'd like to make the changes in the azure_rm_virtualmachine_facts module, with your eventual support, if you don't mind. Will you add a result format arg in the Azure module utils?

Reply:
You don't need to be an AWS module to use the ec2 module utils. Just do the following

```
from ansible.module_utils.ec2 import camel_dict_to_snake_dict

camel_dict_to_snake_dict(vm_info)
```

Reply:
That's fair enough, the code should be split out into a generic function so it can be used by other modules. There is nothing AWS specific about the camel_dict_to_snake_dict and the reverse function, it was only used there because on AWS had a need for it.

Reply:
Yeah, it's only AWS-specific because we only use it in AWS. I'm not sure what to name the new file it would be moved to (`dict_transformations.py`?) but moving it out then re-exposing in the AWS utils for backwards compatibility seems like a good idea.

Reply:
I think dict_transformations.py is a good name :)

Reply:
@zikalino Will you add an output format arg in the Azure module utils? What would each option (raw, default and minimal) look like?

Reply:
@gustavomcarmo please check my pr #39722 , i have 2 options now, raw and curated

Reply:
@zikalino I was making the required changes in order to add the format arg during the weekend, when unfortunately I got the problems I mentioned to you. The fact is that the vms returned by the list function can't be serialized, once bring extensions with no location.